core/src/instance.rs

```diff
@@ -76,54 +76,77 @@ impl Instance {
         )
     }
 
-    /// Start the Event Loop on a seperate thread
-    pub fn start_action_loop(&mut self, context: Arc<Context>) {
+    /// Returns recievers for actions and observers that get added to this instance
+    fn initialize_channels(&mut self) -> (Receiver<ActionWrapper>, Receiver<Observer>) {
         let (tx_action, rx_action) = channel::<ActionWrapper>();
         let (tx_observer, rx_observer) = channel::<Observer>();
         self.action_channel = tx_action.clone();
         self.observer_channel = tx_observer.clone();
 
-        let state_mutex = Arc::clone(&self.state);
+        (rx_action, rx_observer)
+    }
+
+    /// Start the Event Loop on a seperate thread
+    pub fn start_action_loop(&mut self, context: Arc<Context>) {
+        let (rx_action, rx_observer) = self.initialize_channels();
+
+        let sync_self = self.clone();
```

Comment:
I just stumbled across this line in develop and was amazed that this did not break something yet.
Cloning self means that we have two different instances of Holochain, one that is accessible from the outside (the container) and one that holds and mutates the state inside the action loop. That is bad and I think the only reason this could get through tests is that we don't have integration tests yet that actually run an app - exactly what we are adding right now with the app-spec-rust repo and adding holoconsole (and thus a whole container instance) to the CI.

This PR was merged with only one approval. Sorry for having been late on reviewing this, but even if that happens we should not break our rules and merge without sufficient review, @timotree3, @thedavidmeister.

I will change this in the context of https://github.com/holochain/holochain-rust/issues/256 and an upcoming PR. Just FYI.

Reply:
@lucksus it could also be that the original implementation had no tests for consistency of state.

it's ironic because this branch is a step towards making it easier to test for that type of thing by allowing us to step through "the loop" one iteration at a time

actually this is a good lead to chase up re: why tests are randomly broken at the moment!

Reply:
I didn't know about that policy of two reviews. Thanks for letting me know. 

I think that there is a misunderstanding here about what calling `.clone()` on an `Instance` actually does. 
If my understanding is correct (see explanation below), then cloning this `Instance` actually just creates new references to the same structures, and I don't see how it would have a problem with other accesses to the `Instance`.

So here's my explanation. 

The `impl` of `Clone` for `Instance` that is currently provided by `#[derive(Clone)]` is equivalent to this code.
```rust
impl Clone for Instance {
    fn clone(&self) -> Self {
        Instance {
            state: Arc::clone(&self.state), // creates a new reference to the same mutex
            action_channel: self.action_channel.clone(), // creates a new producer for the same channel
            observer_channel: self.observer_channel.clone(), // creates a new producer for the same channel
        }
    }
}
```
The code above (as mentioned in the comments) just creates new accesses to the same underlying concurrent data structures.
The call to `<Instance as Clone>::clone` that you're referring to has the following code:
https://github.com/holochain/holochain-rust/blob/4426c88a89d62dbf0d02136e3eeaef0a4fd11ce6/core/src/instance.rs#L93
If we substitute that `impl` of `Clone` we get this code.
```rust
let sync_self = Instance {
    state: Arc::clone(&self.state),
    action_channel: self.action_channel.clone(),
    observer_channel: self.observer_channel.clone(),
};
```
The reason that we're creating a "new" `Instance` here is so that we can call `process_action` with fewer explicit arguments than if we defined it as:
```rust
fn process_action(
        state: Arc<RwLock<State>>,
        action_channel: Sender<ActionWrapper>,
        observer_channel: Sender<Observer>,
        action_wrapper: ActionWrapper,
        mut state_observers: Vec<Observer>,
        rx_observer: &Receiver<Observer>,
        context: &Arc<Context>,
    ) -> Vec<Observer>
```
Before, we just didn't give the thread an `Instance` and copied the fields of an `Instance` individually and gave them to the thread.
https://github.com/holochain/holochain-rust/blob/5fdf4d11b1155eaaf2eea3777789b873d24926e8/core/src/instance.rs#L83-L86

Thanks for bringing this up. I agree that the usage of `clone` is a bit misleading and we should probably have some comments or change it to the explicit version where we construct a new `Instance` from cloning it's fields.
Does this clarification resolve your concerns or am I missing something, @lucksus @thedavidmeister?

Reply:
i think the hanging tests are due to incorrect history length checking.... again...

Reply:
You are right, @timotree3. Thanks for making this clearly visible! 👍 
I got fooled by my precondition concepts reading that line.. :)